### DISCUSSION TITLE
Wildcard (asterisk) notation in Arch Plugin?

### DISCUSSION THREAD

**Discussion by michael-rubel** [pestphp/pest#728] [Ideas]
## About

It seems that the arch plugin doesn't work with specifying wildcards (*) in the target.

I have different bounded contexts with the same architecture, like `App\BoundedContext\*\Infrastructure\Eloquent\Models` and `App\BoundedContext\*\Infrastructure\Repositories`, so we'll need to hardcode rules for each context individually.

Any plans to address this?

**Comment by fabio-ivona** [comment]
Hi @michael-rubel we are internally evaluating if it's worth to adopt this feature :wink:

**Comment by szabizs** [comment]
I am trying to do also this wildcard stuff using a recursive approach but no success.

```
test('is every form request a request', function () {
    $files = File::allFiles(app_path('Http/Requests'));
    foreach ($files as $file) {
        expect($file->getFilename())->toHaveSuffix('Request.php');
    }

})->group('RequestSuffixTests');
```

`Tests:    1 passed (31 assertions)` I should have one failing file.

I also tried using a `dataset`

```
dataset('requests_files', function () {
    $fileContainer = [];
    foreach (File::allFiles(app_path('Http/Requests')) as $file) {
        $fileContainer[] = $file->getFilename();
    }

    return $fileContainer;
});

test('every form request is a request', function ($file) {
    expect($file)
        ->toHaveSuffix('Request.php');
})->with('requests_files')->group('RequestSuffixTests');
```

But I get this error
![image](https://github.com/pestphp/pest/assets/5064872/bf44d578-317f-4fc7-8b91-316a02c421fd)

I think the issue is how the dataset is transmitted as a parameter to the test.


Later edit:

The same code in PHP Unit is working.

```
public function testThatEveryRequestHasARequestSuffix()
    {
        $fileContainer = [];
        foreach (File::allFiles(app_path('Http/Requests')) as $file) {
            $fileContainer[] = $file->getFilename();
        }

        foreach ($fileContainer as $file) {
            assert(Str::endsWith($file, 'Request.php'), $file);
        }
    }
```

<img width="657" alt="image" src="https://github.com/pestphp/pest/assets/5064872/a1d6ab98-d5d2-4038-954e-64ca80d5ac22">

**Comment by dabekf** [comment]
Wildcards would be a useful feature in a project I'm working on. It is using nWidart/laravel-modules so files are scattered between namespaces, for example `Modules\Homepage\Services`, `Modules\Offers\Services`, `Modules\Autocompleter\Services` etc.

I've added a function to Pest.php which reads all module names from the directory, so the tests look like this:
```php
test('services', function () {
    getModules()->each(
        fn ($module) => expect("Modules\\{$module}\Services")
            ->classes()
            ->toHaveSuffix('Service')
    );
});
```

Wildcard would make them much more concise:
```php
test('services')->expect("Modules\*\Services")->classes()->toHaveSuffix('Service');
```

**Comment by innocenzi** [comment]
I second that, came to the repository to make the suggestion as well. Would be very useful in module-based architectures.

Support for double-wildcard on `uses` as well:
```php
uses(LazilyRefreshDatabase::class, Tests\TestCase::class)->in(
    '../src/Feature/**/Tests/**/*Test.php',
);
```

That would allow me to catch these tests files:
```
src/Feature/OperationsReporting/Tests/OperationsReportingTest.php
src/Feature/OperationsReporting/GetOperationsFromService/Tests/GetOperationsFromService.php
```

**Comment by BernhardK91** [comment]
When using a DDD folder structure, it would be well-needed to allow wildcards in the provided namespace because then the classes are not sorted by "type" but by domain.

F.e. Actions are not located under `App\Actions`, but instead under `App\Domain\*\Actions`.

**Comment by naxvog** [comment]
PR for non-globstar implementation: https://github.com/pestphp/pest-plugin-arch/pull/12
